src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -360,6 +358,53 @@ private static void AddProjectFileDependenciesForPackageReference(PackageSpec pr
             }
         }
 
+        private void AddProjectCentralTransitiveDependencyGroupsForPackageReference(PackageSpec project, LockFile lockFile, IEnumerable<RestoreTargetGraph> targetGraphs)
+        {
+            // TargetFrameworkInformations should have only distinct FrameworkNames
+            var centralPackageVersionsPerFramework = project.TargetFrameworks.ToDictionary(tfmi => tfmi.FrameworkName, tfmi => tfmi.CentralPackageVersions);
+
+            foreach (var targetGraph in targetGraphs.OrderBy(graph => graph.Framework.ToString(), StringComparer.Ordinal))
+            {
+                if (_includeFlagGraphs.TryGetValue(targetGraph, out Dictionary<string, LibraryIncludeFlags> dependenciesIncludeFlags))
+                {
+                    // The transitive dependencies enforced by the central package version management file are written to the assets to be used by the pack task.
+                    IEnumerable<LibraryDependency> centralEnforcedTransitiveDependencies = targetGraph
+                        .Flattened
+                        .Where(graphItem => graphItem.IsCentralTransitive
+                            && centralPackageVersionsPerFramework.ContainsKey(targetGraph.Framework)
+                            && centralPackageVersionsPerFramework[targetGraph.Framework].ContainsKey(graphItem.Key.Name))
+                        .Select((graphItem) =>
+                        {
+                            var matchingCentralVersion = centralPackageVersionsPerFramework[targetGraph.Framework][graphItem.Key.Name];
+                            var libraryDependency = new LibraryDependency()
+                            {
+                                LibraryRange = new LibraryRange(matchingCentralVersion.Name, matchingCentralVersion.VersionRange, LibraryDependencyTarget.Package),
```

Comment:
Should we use the actual selected version instead of the defined one?
This behavior is specific to transitive deps. 

Also we need to account for floating versions in central package versions. Probably in general. Floating versions in transitively should not be allowed. 
I remember discussing this in the design meetings, but I've never seen it accounted for in the code, correct me if I'm wrong. 
The floating part is important, but it can be handled as a future item (by future I mean before the full feature ships).

Reply:
Added issue https://github.com/NuGet/Home/issues/9384 for the floating recommendation. 



> Should we use the actual selected version instead of the defined one?
> This behavior is specific to transitive deps.

I treated these dependencies as the direct dependencies. If we use the resolved version we can have situations when two packs will result in different packages depending on the feed used. I am not sure if this is a concern. 

The PackTask logic has logic to deal with floating versions
https://github.com/NuGet/NuGet.Client/blob/1f50c1d20ede0233cc12dd0213cffae599f99aa4/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs#L871

I see your point as well. I do not feel strong about any of the approaches. I will get another opinion.